### ISSUE TITLE
setoid_rewrite says Invalid_argument("decomp_pointwise")

### ISSUE THREAD

**Issue by coqbot** [coq/coq#4617]
Note: the issue was created automatically with bugzilla2github tool

Original bug ID: BZ#4617
From: @fpottier
Reported version: 8.5
CC:   coq-bugs-redist@lists.gforge.inria.fr

**Comment by andres-erbsen** [comment]
Here is a minimized example (which I found independently and am guessing to be another instance of this bug).

~~~~~
Require Coq.Setoids.Setoid.
Axiom proof_admitted : False. Axiom key value rep : Type. Axiom A D E P : rep -> Prop.
Axiom iff1 : forall {T : Type} (P Q:T->Prop), Prop.
Definition sep (p q : rep -> Prop) (m:rep) := exists (mp:rep), p mp. (* admitting this avoids anomaly *)
Definition exi {T} (P : T -> rep -> Prop) := fun m => exists x:T, P x m. (* admitting this avoids anomaly *)
Axiom sep_exi_l : forall {T} p q, iff1 (sep (@exi T q) p) (exi (fun h => sep (q h) p)). (* changing to (T:=unit) avoids anomaly *)
Goal iff1 (sep (exi (fun h0 : unit => sep A E)) D) P.
try rewrite_strat (outermost sep_exi_l).
try rewrite_strat (bottomup sep_exi_l).
(* Error: Anomaly "Uncaught exception Invalid_argument("decomp_pointwise")." *)
(* Please report at http://coq.inria.fr/bugs/. *)
Abort.
~~~~~~

**Comment by andres-erbsen** [comment]
This is an anomaly, not just a tactic failure. This means triggering it in any proof script causes the entire script to fail without any means of recovering from it. Since I don't see how to avoid triggering this bug from a reusable proof script, it essentially makes rewrite_strat unusable for me.

**Comment by JasonGross** [comment]
Interestingly, this seems to infinite-loop:
```coq
Require Coq.Setoids.Setoid.
Axiom as_prop : forall {T}, T -> Prop.
Axiom A : Prop.
Axiom foo : forall T q, eq (exists (mp : unit) (x : T), q x mp) (exists (x : T) (mp : unit), q x mp).
Goal as_prop (exists (_ : unit) (_ : unit), A).
  try rewrite_strat (bottomup foo).
```

cc @mattam82 @ppedrot

**Comment by JasonGross** [comment]
Here is an even more minimized example with the same anomaly:
```coq
Require Coq.Setoids.Setoid.
Axiom as_prop : forall {T}, T -> Prop.
Axiom foo : forall T, (fun _ : unit => T /\ True) = (fun _ : unit => T).
Goal as_prop (fun _ : unit => True /\ True).
  try rewrite_strat (bottomup foo).
```

**Comment by andres-erbsen-sifive** [comment]
Another example:

```coq
Require Import Eqdep Setoid.
Goal forall (t : unit) (pf : tt = t) S (P : S t -> unit),
  match pf in _ = _ with eq_refl => fun _ => tt end = P.
Proof.
  intros.
  try setoid_rewrite <-Eqdep.Eq_rect_eq.eq_rect_eq.
Abort.
```

**Comment by mattam82** [comment]
The decomp_pointwise anomaly happens because rewriting (plugins/ltac/rewrite.ml:L289) tries to decompose a relation constraint and does not find a `pointwise_relation A B R` there that should be produced when rewriting under a binder of type `A` using relation `R`. Probably the constraint it tries to decompose is (?R ==> impl) as it rewrites at the top of the goal. I guess the fix is to be more flexible and generate a subrelation constraint of the form `pointwise_relation ?A ?B ?S <= ?R ==> impl` instead of forcing a syntactic relation constraint.

**Comment by ppedrot** [comment]
The examples from this issue seem to have been fixed, but weirdly enough #11347 is not.

**Comment by SkySkimmer** [comment]
>The examples from this issue seem to have been fixed, but weirdly enough #11347 is not.

The examples here use `try`, if you remove it you still see the anomaly on some.

I guess at some point setoid_rewrite started putting the invalid_argument in tclzero (between 8.11 and 8.12)
